google-logging-utils/lib/google/logging/message.rb

```diff
@@ -0,0 +1,318 @@
+          source_location: @source_location,
+          insert_id: @insert_id,
+          trace: @trace,
+          span_id: @span_id,
+          trace_sampled: @trace_sampled,
+          labels: @labels.dup
+        }
+      end
+
+      # @private
+      def == other
+        return false unless other.is_a? Message
+        message == other.message &&
+          fields == other.fields &&
+          timestamp == other.timestamp &&
+          source_location == other.source_location &&
+          insert_id == other.insert_id &&
+          trace == other.trace &&
+          span_id == other.span_id &&
+          trace_sampled? == other.trace_sampled? &&
+          labels == other.labels
+      end
+      alias eql? ==
+
+      # @private
+      def hash
+        [message, fields, timestamp, source_location, insert_id, trace, span_id, trace_sampled, labels].hash
+      end
+
+      private
+
+      # @private
+      DISALLOWED_FIELDS = [
+        "severity",
+        "message",
+        "log",
+        "httpRequest",
+        "timestamp"
+      ].freeze
+
+      def interpret_fields fields
+        return nil if fields.nil?
+        fields = normalize_json fields
+        fields.each_key do |key|
+          if DISALLOWED_FIELDS.include?(key) || key.start_with?("logging.googleapis.com/")
+            raise ArgumentError, "Field key not allowed: #{key}"
+          end
+        end
+        fields
+      end
+
+      def normalize_json input
+        case input
+        when Hash
+          input.to_h { |k, v| [k.to_s, normalize_json(v)] }
+        when Array
+          input.map { |v| normalize_json v }
+        when Integer, Float, nil, true, false
+          input
+        when Rational
+          if input.denominator == 1
+            input.numerator
+          else
+            input.to_f
+          end
+        else
+          input.to_s
+        end
+      end
+
+      def interpret_message message, fields
+        if message
+          message = message.to_s
+          full_message = fields ? "#{message} -- #{JSON.generate fields}" : message
+          [message, full_message]
+        else
+          fields_json = JSON.generate fields
+          [fields_json, fields_json]
+        end
+      end
+
+      def interpret_timestamp timestamp
+        case timestamp
+        when Time
+          timestamp
+        when Numeric
+          Time.at timestamp
+        when :now
+          Time.now.utc
+        end
+      end
+
+      def interpret_source_location source_location
+        case source_location
+        when SourceLocation
+          source_location
+        when Hash
+          SourceLocation.new(**source_location)
+        when :caller
+          SourceLocation.for_caller omit_files: [__FILE__]
+        end
+      end
+
+      def interpret_string input
+        return nil if input.nil?
+        input.to_s
+      end
+
+      def interpret_boolean input
+        return nil if input.nil?
+        input ? true : false
```

Comment:
I know there are differing opinions on this (and rubocop provides [different options](https://docs.rubocop.org/rubocop/cops_style.html#styledoublenegation)), but personally I dislike the `!!` idiom because it doesn't "say" what it "means".